libraries/standard/mqtt/src/mqtt.c

```diff
@@ -41,6 +41,110 @@ static int32_t sendPacket( MQTTContext_t * pContext,
                            const uint8_t * pBufferToSend,
                            size_t bytesToSend );
 
+/**
+ * @brief Calculate the interval between two timestamps, including when the
+ * later value has overflowed.
+ *
+ * @param[in] later The later time stamp.
+ * @param[in] start The earlier time stamp.
+ *
+ * @return later - start.
+ */
+static uint32_t calculateElapsedTime( uint32_t later, uint32_t start );
+
+/**
+ * @brief Convert a byte indicating a publish ack type to an #MQTTPubAckType_t.
+ *
+ * @param[in] packetType First byte of fixed header.
+ *
+ * @return Type of ack.
+ */
+static MQTTPubAckType_t getAckFromPacketType( uint8_t packetType );
+
+/**
+ * @brief Receive bytes into the network buffer, with a timeout.
+ *
+ * @param[in] pContext Initialized MQTT Context.
+ * @param[in] bytesToRecv Number of bytes to receive.
+ * @param[in] timeoutMs Time remaining to receive the packet.
+ *
+ * @return Number of bytes received, or -1 on network error.
```

Comment:
The transport read function returns "a negative number" if it encounters an error. I'm not sure if we will have different error codes but I can update the log to have that return value.

Reply:
We can potentially say value less than zero, rather than -1 to be aligned with transport layer spec.

Reply:
Actually it was -1, because I based this off the original implementation of recvExact we had that returned -1. I changed it to bubble up the exact error code.